hadoop-tools/hadoop-azure/src/test/java/org/apache/hadoop/fs/azurebfs/ITestAbfsReadFooterMetrics.java

```diff
@@ -0,0 +1,268 @@
+
+public class ITestAbfsReadFooterMetrics extends AbstractAbfsScaleTest {
+
+  public ITestAbfsReadFooterMetrics() throws Exception {
+  }
+
+  private static final String TEST_PATH = "/testfile";
+
+  @Test
+  public void testReadFooterMetricsWithParquetAndNonParquet() throws Exception {
+    testReadWriteAndSeek(8 * ONE_MB, DEFAULT_READ_BUFFER_SIZE, ONE_KB, 4 * ONE_KB);
+  }
+
+  @Test
+  public void testReadFooterMetrics() throws Exception {
+    int bufferSize = MIN_BUFFER_SIZE;
+    final AzureBlobFileSystem fs = getFileSystem();
+    final AbfsConfiguration abfsConfiguration = fs.getAbfsStore()
+        .getAbfsConfiguration();
+    abfsConfiguration.set(FS_AZURE_METRIC_FORMAT, String.valueOf(MetricFormat.INTERNAL_FOOTER_METRIC_FORMAT));
+    abfsConfiguration.set(FS_AZURE_TRACINGMETRICHEADER_FORMAT,
+        String.valueOf(TracingHeaderFormat.INTERNAL_FOOTER_METRIC_FORMAT));
+    abfsConfiguration.setWriteBufferSize(bufferSize);
+    abfsConfiguration.setReadBufferSize(bufferSize);
+
+    final byte[] b = new byte[2 * bufferSize];
+    new Random().nextBytes(b);
+
+    Path testPath = path(TEST_PATH);
+    FSDataOutputStream stream = fs.create(testPath);
+    try {
+      stream.write(b);
+    } finally {
+      stream.close();
+    }
+    IOStatisticsLogging.logIOStatisticsAtLevel(LOG,
+        IOSTATISTICS_LOGGING_LEVEL_INFO, stream);
+
+    final byte[] readBuffer = new byte[2 * bufferSize];
+    int result;
+    IOStatisticsSource statisticsSource = null;
+    try (FSDataInputStream inputStream = fs.open(testPath)) {
+      statisticsSource = inputStream;
+      ((AbfsInputStream) inputStream.getWrappedStream()).registerListener(
+          new TracingHeaderValidator(abfsConfiguration.getClientCorrelationId(),
+              fs.getFileSystemId(), FSOperationType.READ, true, 0,
+              ((AbfsInputStream) inputStream.getWrappedStream())
+                  .getStreamID()));
+      inputStream.seek(bufferSize);
+      result = inputStream.read(readBuffer, bufferSize, bufferSize);
+      assertNotEquals(-1, result);
+
+      //to test tracingHeader for case with bypassReadAhead == true
+      inputStream.seek(0);
+      byte[] temp = new byte[5];
+      int t = inputStream.read(temp, 0, 1);
+
+      inputStream.seek(0);
+      result = inputStream.read(readBuffer, 0, bufferSize);
+    }
+    IOStatisticsLogging.logIOStatisticsAtLevel(LOG,
+        IOSTATISTICS_LOGGING_LEVEL_INFO, statisticsSource);
+
+    assertNotEquals("data read in final read()", -1, result);
+    assertArrayEquals(readBuffer, b);
+    AbfsReadFooterMetrics abfsReadFooterMetrics = fs.getAbfsClient()
+        .getAbfsCounters()
+        .getAbfsReadFooterMetrics();
+    String readFooterMetric = "";
+    if (abfsReadFooterMetrics != null) {
+      readFooterMetric = abfsReadFooterMetrics.toString();
+    }
+    assertEquals(
+        "$NonParquet:$FR=16384.000_16384.000$SR=1.000_16384.000$FL=32768.000$RL=16384.000",
```

Comment:
There is a grammar behind this string. Can we have a grammar ingestor which ingests values to create a string, and that string is asserted. Also, in the production code as well if can have the same grammar ingestor in the `toString()`. Benefit of that shall be that production code and test code string creation shall be in sync always.